README.md
# powershell-hsdp-iam

[![CI](https://github.com/philips-software/powershell-hsdp-iam/workflows/CI/badge.svg)](https://github.com/philips-software/powershell-hsdp-iam/actions?query=workflow%3ACI)
[![powershellgallery](https://img.shields.io/powershellgallery/v/hsdp-iam.svg)](https://www.powershellgallery.com/packages/hsdp-iam)

[Visit HSDP Slack Support](https://hsdp.slack.com)

**Description**: This powershell module contains cmdlets to work against the HSDP IAM APIs using a powershell approach.

  - **Technology stack**: Microsoft Powershell. Compatible with PS Core.
  - **Key concepts** powershell  (technical, philosophical, or both) important to the user’s understanding.
  - **Status**:  Alpha. See ToDos. [CHANGELOG](CHANGELOG.md).

## Status

The intent of this module is to cover all API functionality available in the HSDP IAM APIs.

The coverage is tracked in [coverage.md](coverage.md)

## Dependencies

* Powershell 7/Core
* Pester for unit tests

## Installation

This package is available on [the powershell gallery](https://www.powershellgallery.com/packages/hsdp-iam)

```
Install-Module hsdp-iam
Import-Module hsdp-iam
```

## Configuration
A new configuration file may be created using the New-Config cmdlet. The cmdlet will prompt for the following configuration parameters:

| Value             | Example | Purpose | Notes|
|-------------------|---------|---|----|
| HSDP IAM URL | https://iam-integration.us-east.philips-healthsuite.com | IAM base URI | Do not include trainling slash
| HSDP IDM URL | https://idm-integration.us-east.philips-healthsuite.com   | IDM base URI | Do not include trainling slash
| HSDP Username and Password |  | Credentials with org administrative role | always required
| HSDP ClientId and Secret | | IAM client key and secret | always required
| HSDP App Shared Key and Secret | | App shared key and secret  | required for API Signature headers

The configuration will be saved to a file named ```./config.xml``` The secrets provided will be encrypted the configuration when using the windows platform.

## Import configuration
To authenticate using the configuration execute the following cmdlet:

```powershell
Set-FileConfig
```

## Concepts:

These powershell cmdlets have been designed to follow common powershell patterns:

### Online Help

The cmdlets all contain online help which can be obtains by using the Get-Help cmdlet. For instance:

```powershell
Get-Help Get-Org
```

### Objects
The cmdlets in this module return PSObjects from most functions that match the HSDP models. So for example:

```powershell
$org = Get-Org "02bdfa45-db4b-4450-a77e-b59ab9df9472"
```

The $Org will be represented in the following PSObject:

```
schemas           : {urn:ietf:params:scim:schemas:core:philips:hsdp:2.0:Organization}
id                : d578177f-f3db-4919-805a-b382c6fa0032
name              : _testorg1
parent            : @{value=e5550a19-b6d9-4a9b-ac3c-10ba817776d4; $ref=https://idm-integration.us-east.philips-healthsuite.com/authorize/scim/v2/Organizations/e5550a19-b6d9-4a9b-ac3c-10ba817776d4}
active            : True
inheritProperties : True
owners            : {@{value=36268f6b-828e-454c-aba8-8c2044fb19f9; primary=True}}
createdBy         : @{value=36268f6b-828e-454c-aba8-8c2044fb19f9}
modifiedBy        : @{value=36268f6b-828e-454c-aba8-8c2044fb19f9}
meta              : @{resourceType=Organization; created=2/17/2020 6:05:25 PM; lastModified=2/17/2020 8:36:25 PM; location=https://idm-integration.us-east.philips-healthsuite.com/authorize/scim/v2/Organizations/d578177f-f3db-4919-805a-b382c6fa0032;
                    version=W/"617601800"}
```

The PSObjects are used by other CmdLets to simplify parameter passing. For instance:

```powershell
$groups = Get-Groups -Org $org
```

Will return all the groups in the org object passed to the cmdlet.

### Object Update (version meta)
Cmdlets that perform updates will leverage the $object.meta.version property to pass the proper ETag.

### Pipeline support

Many cmdlets use ValueFromPipeline allowing composition such as in the following example to remove all users from an org:

```powershell
Get-UserIds $org | Remove-User
```

### Debugging and Tracing

The cmdlets have all be instrumented with Debug and Verbose information. This can be enabled using:

```powershell
$DebugPreference="continue"
$VerbosePreference="continue"
```

## Cmdlet Categories
- Configuration
- Application
- Client
- OAuth2
- Group
- Organization
- Proposition
- Role
- Service
- User
- Utility

## Recepies

### Add Org and User

```powershell
$parentOrg = Get-Org "02bdfa45-db4b-4450-a77e-b59ab9df9472"
$org = Add-Org -ParentOrg $parentOrg -Name "MyNewOrg"
$user = Add-User -LoginId "test01" -Email "asdfasdf@mailinator.com" -MobilePhone "1234512345" -FamilyName "FAMILY" -GivenName "GIVEN" -Org $org
```

### Create Proposition, Application, Service Identity and generate a JWT

```powershell
$org = Get-Org "02bdfa45-db4b-4450-a77e-b59ab9df9472"
$propId = ([GUID]::NewGuid().ToString('D'))
$prop = Add-Proposition -Org $org -Name "My Proposition" -GlobalReferenceId $propId

$appId = ([GUID]::NewGuid().ToString('D'))
$app = Add-Application -Proposition $prop -Name "My Application" -GlobalReferenceId $appId

$keyFile = "$($appName).pem"
# create a new service and write a key file
$service = Add-Service -Application $app -Name "My Service" -PrivateKeyPath $keyFile

$jwt = New-HsdpJWT -Service $service -KeyFile $keyFile
```

### Assign multiple users into group

```powershell
$org = Get-Org "02bdfa45-db4b-4450-a77e-b59ab9df9472"
Set-UsersInGroup -Org $Org -GroupName "My Group" -UserIds @("user1@mailinator.com", "user2@mailinator.com")
```

### Find a user by email address in an org and display all the permissions for all the roles

```powershell
$user = (Get-UserIds -Org $org | Get-User | Where-Object { $_.emailAddress -eq "user1@mailinator.com" })
$user.memberships.roles | % { get-roles -Name $_ } | get-permissions
```

### Display all the email addresses of users who have never verified their email across all organizations

```powershell
Get-Orgs | Get-UserIds | Get-User | Where-Object { $_.accountStatus.emailVerified -ne "True" } | Select-Object -ExpandProperty emailAddress
```
## How to test the software

### Unit Tests
Complete unit tests are written using Pester Version 5

The ```hsdp-iam.tests.ps1``` will execute all unit tests. This is used as part of the CD/CD verification pipeline as well.

### Sanity Tests
A sanity test in ```Sanity.Tests.ps1``` will execute as part of the build verification.

Sanity test can be run locally but all configuration must be passed to the script. (e.g. full HSDP IAM env configuration)

### Integration Tests
Integration tests are integrated into the build pipeline and must all pass on pull requests.

## Known issues

None

## Contributing
See [CONTRIBUTE.md](CONTRIBUTE.md)


## Contact / Getting help

mark.lindell@philips.com

## License

Link to [LICENSE.md](LICENSE.md)

## Credits and references

1. Inspiration for powershell approach taken from from [JiraPS](https://github.com/AtlassianPS/JiraPS)

[![Slack](https://philips-software-slackin.now.sh/badge.svg)](https://philips-software-slackin.now.sh)


CHANGELOG.md
## 0.1.35

* Add-User: Throws exception on anything but 201 response (previously did not throw on 200)

## 0.1.34

* ``New-Config`` now supports value from pipeline

``` powershell
New-Config @p -Prompt $false -Path $null | Set-Config
```

## 0.1.33

* Added support for v2 API in `Get-Evaulate`
* Updated `New-Config` to support not creating a configuration file. Allows scripts to define and configure inline do scripts can be self contained.

``` powershell
$p = @{
    IamUrl = ""
    IdmUrl = ""
    CredentialsUserName = ""
    CredentialsPassword = ""
    ClientCredentialsUserName = ""
    ClientCredentialsPassword = ""
    AppCredentialsUserName = ""
    AppCredentialsPassword = ""
}
Set-Config (New-Config @p -Prompt $false -Path $null)
```
*NOTE: Using this approach will put your passwords/secrets in plain text in a script file. Use with caution!*

## 0.1.32

Fixed bug with `Add-AuthPolicy`

## 0.1.31

Added  support for ​/authorize​/access​/Policy API:
* `Add-AuthPolicy`
* `Get-AuthPolicy`
* `Get-AuthPolicies`
* `Remove-AuthPolicy`

## 0.1.29

* Move distribution to Powershell Gallary.

## 0.1.27

* Added cmdlet Set-AppServiceScope
* Update contribute for coding style and best practices reference

## 0.1.26

* Added cmdlets to support Device API

## 0.1.25

* Added support to use refresh token to retrieve new access token if the access token expires
* Re-worked how authentication value is stored between cmdlets
* Renamed `Get-Headers` to `Get-AuthenticationHeader` - changed return value to string that just contains value for the authenticate header  rather than a hashtable of a subset of header values.
* `Set-Config` now clears cached authentication and re-authenticates

## 0.1.24 - Add Cmdlets for PasswordPolicy

Added Cmdlets:
* `Add-PasswordPolicy`
* `Get-PasswordPolicies`
* `Get-PasswordPolicy`
* `Set-PasswordPolicy`
* `Remove-PasswordPolicy`
* `Get-UserKba`
* `Reset-UserPassword`

## 0.1.23 - Updates for HSDP May 2020 release

Changed documentation and params to support batches of 100 for:
* `Add-Group`
* `Set-GroupRole`
* `Add-Permissions`

Changed `Add-Permissions` param -Role to -Roles to support multiple

## 0.1.22 - Added more user cmdlets

Renamed ``Get-Users`` cmdlet to ``Get-UserIds`` for clarity. Get-UserIds now support paging

New Cmdlets added:
* `New-UserPassword`
* `New-UserResendActivation`
* `Reset-UserMfa`
* `Set-UserKba`
* `Set-UserNewLoginId`
* `Set-UserPassword`
* `Set-UserUnlock`

## 0.1.21 - Added new cmdlets for OAuth2 Client API.

Configuration changes:
* Removed unused "OAuth2Credentials" config params
* Support for scopes has now been added to the CmdLets New-Config and Read-Config to support the scopes requested when performorming.
* The default scopes requested of `/authorize/oauth2/token` is `@("profile","email","read_write")` unless otherwise specified

The following CmdLets have been renamed, support pagination for querying multiple resources, and support all query parameters of the API
* `Get-Proposition` renamed to `Get-Propositions`
* `Get-Application` renamed to `Get-Applications`

Improvements:
* Established and configured a new HSDP tenant dedicated to this project.
* Integration tests now part of build pipeline
* Minor inline documentation updates

## 0.1.19 - Unit Testing Coverage Complete

## 0.1.1 - Version build & published from github action

## 0.0.1 - Inital release



Public/User/Add-User.ps1
<#
    .SYNOPSIS
    Register a user account.

    .DESCRIPTION
    Creates a new user within an organization. A user can be registered in the system in two ways
    
    Admin Managed
    A user with 'USER.WRITE' permission can register users within an organization on behalf of the user. 
    In this case, managingOrganization is mandatory and user will be created under specified organization. 
    The registration triggers an account activation email to the registered email address of the user. 
    Upon activation, the user shall set the password for the account and update the complete user profile.

    Self Managed
    Consumer applications can trigger user registration in a way that does not need an organization. 
    All self-managed users will be created under a system organization called Consumer organization. 
    Self-registration sends an account activation email to the registered email address of the user. 
    Upon activation, the user shall set the password for the account and update the complete user profile. 
    Self managed registration request must set isAgeValidated property to true.

    Minimum properties required to create a user are loginId, emailAddress, givenName and familyName.

    .INPUTS
    An organization PSObject

    .OUTPUTS
    A user PSObject

    .PARAMETER Org
    An organization PSObject

    .PARAMETER LoginId
    Unique login ID for the user.
    pattern: ^((?![~`!#%^&*()+={}[\]|/\\<>,;:"'?])[\S])*$

    .PARAMETER Email
    Email ID
    Allowed characters are _A-Za-z0-9-@ with a minimum length of 3, max of 255.
    
    The sequence of characters of email address should be as follows:
        a. Any character one or more times.[ _ A to Z a to z 0 to 9 to +]
        b. If there is a '.' then it should also be followed by one or more characters [ _ A to Z a to z 0 to 9 -]. A '.'  without characters following it is not allowed 
        c. '@' is required
        d. Any character[ A to Z a to z 0 to 9 -] one or more times
        e. If there is a '.' then it should also be followed by one or more characters [A to Z a to z 0 to 9].
        f. A '.' without characters following it is not allowed A '.' followed by any alphabet 2 or more times

    .PARAMETER LoginId    
    Unique login ID for the user. 
    pattern: ^((?![~`!#%^&*()+={}[\]|/\\<>,;:"'?])[\S])*$

    .PARAMETER MobilePhone
    Mobile number
    pattern: ^([0-9]{10,20})$
    Allowed 0-9 with a minimum length of 10, max of 20.
    Ex: 1234512345

    .PARAMETER GivenName
    Given names (not always 'first'). Includes middle names. pattern: ^((?![0-9,#<>~&^%?*|/\\{}[\]!@$():;+=''"])[\S])*$

    .PARAMETER FamilyName    
    Family name (often called 'Surname'). pattern: ^((?![0-9,#<>~&^%?*|/\\{}[\]/!@$():;+=''"])[\S])*$

    .EXAMPLE
    $user = Add-User -LoginId "test01" -Email "asdfasdf@mailinator.com" -MobilePhone "1234512345" -FamilyName "FAMILY" -GivenName "GIVEN" -ManagingOrgId "e5550a19-b6d9-4a9b-ac3c-10ba817776d4"

    .LINK
    https://www.hsdp.io/documentation/identity-and-access-management-iam/api-documents/resource-reference-api/user-api-v2#/User%20Identity/post_authorize_identity_User

    .NOTES
    POST: /authorize/identity/User v2
#>
function Add-User {

    [CmdletBinding()]
    [OutputType([PSObject])]
    param(
        [Parameter(Mandatory, Position = 0, ValueFromPipeline)]
        [ValidateNotNullOrEmpty()]
        [String]$Org,

        [Parameter(Mandatory, Position = 1)]
        [ValidateNotNullOrEmpty()]
        [String]$LoginId,

        [Parameter(Mandatory, Position = 2)]
        [ValidateNotNullOrEmpty()]
        [String]$Email,

        [Parameter(Mandatory, Position = 3)]
        [ValidateNotNullOrEmpty()]
        [String]$MobilePhone,

        [Parameter(Mandatory, Position = 4)]
        [ValidateNotNullOrEmpty()]
        [String]$GivenName,

        [Parameter(Mandatory, Position = 5)]
        [ValidateNotNullOrEmpty()]
        [String]$FamilyName
    )
     
    begin {
        Write-Verbose "[$($MyInvocation.MyCommand.Name)] Function started"
    }

    process {
        Write-Debug "[$($MyInvocation.MyCommand.Name)] PSBoundParameters: $($PSBoundParameters | Out-String)"


        $body = @{
            "resourceType"= "Person";
            "loginId"= $LoginId;
            "managingOrganization"= $Org.id;
            "name" = @{
                "family"= $FamilyName;
                "given"= $GivenName;
            };
            "telecom" = @(
                @{
                    "system"="mobile";
                    "value"=$MobilePhone;
                },
                @{
                    "system"="email";
                    "value"=$Email;
                }
            );
        }

        $response  = (Invoke-ApiRequest -Path "/authorize/identity/User" -Version 2 -Method Post -Body $body `
            -AddHsdpApiSignature -ValidStatusCodes @(200,201))
        Write-Output @($response)
    }

    end {
        Write-Verbose "[$($MyInvocation.MyCommand.Name)] Complete"
    }   
}